stackslib/src/chainstate/stacks/tests/block_construction.rs

```diff
@@ -5087,3 +5087,188 @@ fn paramaterized_mempool_walk_test(
+            "INSERT INTO nonces (address, nonce) VALUES (?, ?), (?, ?), (?, ?)",
+            params![address_0, 2, address_1, 7, address_2, 9],
+        )
+        .unwrap();
+    mempool_tx.commit().unwrap();
+
+    // Visit transactions. Keep a record of the order of visited txs so we can compare at the end.
+    let mut considered_txs = vec![];
+    let deadline = get_epoch_time_ms() + 30000;
+    chainstate.with_read_only_clarity_tx(
+        &TEST_BURN_STATE_DB,
+        &StacksBlockHeader::make_index_block_hash(&b_2.0, &b_2.1),
+        |clarity_conn| {
+            // When the candidate cache fills, one pass cannot process all transactions
+            loop {
+                if mempool
+                    .iterate_candidates::<_, ChainstateError, _>(
+                        clarity_conn,
+                        &mut tx_events,
+                        MemPoolWalkSettings::default(),
+                        |_, available_tx, _| {
+                            considered_txs.push((
+                                available_tx.tx.metadata.origin_address.to_string(),
+                                available_tx.tx.metadata.origin_nonce,
+                            ));
+                            Ok(Some(
+                                // Generate any success result
+                                TransactionResult::success(
+                                    &available_tx.tx.tx,
+                                    available_tx.tx.metadata.tx_fee,
+                                    StacksTransactionReceipt::from_stx_transfer(
+                                        available_tx.tx.tx.clone(),
+                                        vec![],
+                                        Value::okay(Value::Bool(true)).unwrap(),
+                                        ExecutionCost::zero(),
+                                    ),
+                                )
+                                .convert_to_event(),
+                            ))
+                        },
+                    )
+                    .unwrap()
+                    .0
+                    == 0
+                {
+                    break;
+                }
+                assert!(get_epoch_time_ms() < deadline, "test timed out");
+            }
+            assert_eq!(
+                considered_txs,
+                vec![
+                    (address_0.clone(), 2),
+                    (address_0.clone(), 3),
+                    (address_0.clone(), 4),
+                    (address_0.clone(), 5),
+                    (address_0.clone(), 6),
+                    (address_1.clone(), 7), // Higher fee for address 1
+                    (address_0.clone(), 7),
+                    (address_1.clone(), 8),
+                    (address_0.clone(), 8),
+                    (address_2.clone(), 9), // Higher fee for address 2
+                    (address_1.clone(), 9),
+                    (address_0.clone(), 9),
+                ],
```

Comment:
It looks like the ideal ordering would be:
```
                vec![
                     (address_2.clone(), 9), // Higher fee for address 2
                     (address_1.clone(), 7), // Higher fee for address 1
                     (address_1.clone(), 8),
                     (address_1.clone(), 9),
                     (address_0.clone(), 2),
                     (address_0.clone(), 3),
                     (address_0.clone(), 4),
                     (address_0.clone(), 5),
                     (address_0.clone(), 6),
                     (address_0.clone(), 7),
                     (address_0.clone(), 8),
                     (address_0.clone(), 9),
                 ],
```
We don't want to penalize ready transactions with higher fees just for having a higher nonce. We want to select all of the transactions with the ready nonce, then order that set by fee rate.

Reply:
Great catch, will make that change

Reply:
Great test!

Reply:
Done @obycode , I had a bug in the WHERE condition for the LEFT JOIN

Reply:
Nice. I see how this latest version is basically grouped in sets of nonces that are ready for each round:
- Round 1: (addr 2, nonce 9), (addr 1, nonce 7), (addr 0, nonce 2)
- Round 2: (addr 1, nonce 8), (addr 0, nonce 3)
- Round 3: (addr 1, nonce 9), addr 0, nonce 4)
- Round 4: (addr 0, nonce 5)
- ...

I think this is fine, but is there any good way to handle the chained nonces better? That could be part of a separate PR after this one if there is not a clear solution.

It would be great if we could get a good benchmark setup to compare how changes affect the transaction selection timing.

Reply:
If we push that to a separate PR, can you add a comment in this test specifying the optimal ordering I noted earlier, so we can easily come back to this later.

Reply:
I see what you mean, let me do some tests here and come back to you with an answer

Reply:
The new query fixes this issue and always visits the top ranked transaction first 👍